File: klinara-ios/klinara-ios/Features/Packages/SellPackageSheet.swift
import SwiftUI

/// Paket satışı: tanım seç → önizle → sat.
///
/// **Idempotency anahtarı sheet açılışında üretilir ve satış boyunca sabit
/// kalır.** Ağ hatasından sonra kullanıcının tekrar "Sat"a basması çok olası
/// bir senaryo; her denemede yeni anahtar üretilseydi müşteri iki paket
/// satın almış olurdu.
struct SellPackageSheet: View {

    let session: AppSession
    let store: CustomerPackagesStore

    @Environment(\.dismiss) private var dismiss

    @State private var selectedId: String?
    @State private var note = ""
    @State private var error: APIError?
    /// Sheet ömrü boyunca tek anahtar — `@State` olması kasıtlı.
    @State private var idempotencyKey = UUID().uuidString

    private var definitionStore: PackageDefinitionStore { session.packageDefinitionStore }

    private var options: [PackageDefinition] {
        definitionStore.sellable(in: session.selectedBranchId)
            .sorted { $0.name.localizedStandardCompare($1.name) == .orderedAscending }
    }

    private var selected: PackageDefinition? {
        options.first { $0.id == selectedId }
    }

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: KlinaraMetrics.lg) {
                    if let error, !error.isFieldScoped {
                        ErrorBanner(error: error)
                    }

                    switch definitionStore.state {
                    case .loading:
                        ProgressView()
                            .tint(KlinaraColor.sage)
                            .frame(maxWidth: .infinity)
                            .padding(.vertical, KlinaraMetrics.xl)

                    case .failed(let failure):
                        ErrorBanner(error: failure, onRetry: { Task { await definitionStore.reload() } })

                    case .loaded:
                        if options.isEmpty {
                            EmptyStateView(
                                icon: "shippingbox",
                                title: "Satılabilir paket yok",
                                message: "Bu şubede satışa açık paket tanımı bulunmuyor."
                            )
                        } else {
                            picker
                            if let selected { preview(selected) }
                            noteCard
                            sellButton
                        }
                    }
                }
                .padding(.horizontal, KlinaraMetrics.screenInset)
                .padding(.vertical, KlinaraMetrics.lg)
            }
            .background(KlinaraColor.surface)
            .navigationTitle("Paket sat")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .topBarLeading) {
                    Button("Vazgeç") { dismiss() }
                        .klinaraText(.bodyM)
                        .foregroundStyle(KlinaraColor.charcoalMuted)
                }
            }
            .task { await definitionStore.load() }
            .overlay {
                if store.isSaving { AuthLoadingOverlay(message: "Satış kaydediliyor…") }
            }
        }
        .tint(KlinaraColor.sage)
    }

    // MARK: Bölümler

    private var picker: some View {
        KlinaraSearchablePicker(
            title: "Paket",
            options: options,
            label: { $0.name },
            detail: { "\($0.totalSessions) seans · \(Money.format(minor: $0.totalPriceMinor))" },
            isSelected: { $0.id == selectedId },
            onSelect: { selectedId = $0.id },
            searchPrompt: "Paket ara",
            emptyMessage: "Aramanızla eşleşen paket yok."
        )
    }

    /// Satışta ne olacağı **satmadan önce** görünmeli: kalemler, süre ve
    /// devredilebilirlik satış anında dondurulur, sonradan düzeltilemez.
    private func preview(_ definition: PackageDefinition) -> some View {
        KlinaraCard(title: "Satış önizlemesi", footnote: previewFootnote(definition)) {
            ForEach(Array(definition.items.sorted { $0.sortOrder < $1.sortOrder }.enumerated()), id: \.element.id) { index, item in
                if index > 0 { KlinaraDivider() }
                KlinaraRow(
                    label: item.serviceName,
                    value: "\(item.quantity) seans",
                    detail: Money.format(minor: item.unitListPriceMinor) + " birim"
                )
            }
            KlinaraDivider()
            KlinaraRow(
                label: "Satış tutarı",
                value: Money.format(minor: definition.totalPriceMinor),
                detail: definition.discountMinor.map {
                    "Liste: \(Money.format(minor: definition.listPriceMinor)) · İndirim: \(Money.format(minor: $0))"
                }
            )
        }
    }

    private func previewFootnote(_ definition: PackageDefinition) -> String {
        var parts: [String] = []
        parts.append(definition.validityDays.map { "Satıştan itibaren \($0) gün geçerli" }
            ?? "Süresiz")
        parts.append(definition.isTransferable ? "Devredilebilir" : "Devredilemez")
        parts.append("Satış anındaki fiyat ve kalemler dondurulur; tanım sonradan değişse bile bu paket etkilenmez.")
        return parts.joined(separator: " · ")
    }

    private var noteCard: some View {
        KlinaraCard(title: "Not") {
            KlinaraTextEditor(
                label: "Satış notu",
                text: $note,
                placeholder: "İsteğe bağlı",
                error: error?.fieldErrors["note"],
                minHeight: 80
            )
            .padding(KlinaraMetrics.md)
        }
    }

    private var sellButton: some View {
        KlinaraButton(
            title: "Paketi sat",
            kind: .primary,
            icon: "cart.badge.plus",
            isLoading: store.isSaving,
            isEnabled: selectedId != nil && !store.isSaving
        ) {
            Task { await sell() }
        }
    }

    // MARK: Eylem

    private func sell() async {
        guard let definitionId = selectedId else { return }
        error = nil
        do {
            let trimmed = note.trimmingCharacters(in: .whitespacesAndNewlines)
            _ = try await store.sell(
                definitionId: definitionId,
                note: trimmed.isEmpty ? nil : trimmed,
                idempotencyKey: idempotencyKey
            )
            dismiss()
        } catch {
            self.error = error as? APIError ?? .network
        }
    }
}

File: klinara-android/app/src/main/java/com/klinara/android/features/packages/SellPackageSheet.kt
package com.klinara.android.features.packages

import androidx.compose.foundation.layout.Box
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ShoppingCart
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import com.klinara.android.designsystem.KlinaraTheme
import com.klinara.android.designsystem.KlinaraType
import com.klinara.android.designsystem.components.AuthLoadingOverlay
import com.klinara.android.designsystem.components.EmptyStateView
import com.klinara.android.designsystem.components.ErrorBanner
import com.klinara.android.designsystem.components.KlinaraButton
import com.klinara.android.designsystem.components.KlinaraCard
import com.klinara.android.designsystem.components.KlinaraDivider
import com.klinara.android.designsystem.components.KlinaraRow
import com.klinara.android.designsystem.components.KlinaraScreen
import com.klinara.android.designsystem.components.KlinaraSearchablePicker
import com.klinara.android.designsystem.components.KlinaraTextEditor
import com.klinara.android.services.formatting.Money
import com.klinara.android.services.networking.Loadable
import com.klinara.android.services.packages.PackageDefinition

/**
 * Paket satış sayfası — sheet DEĞİL, gerçek bir hedef (§5.3 Kural 2).
 *
 * Satışta ne olacağı **satmadan önce** görünür: kalemler, süre ve devredilebilirlik satış
 * anında DONDURULUR ve sonradan düzeltilemez.
 */
@Composable
fun SellPackageSheet(
    state: SellPackageUiState,
    branchName: String?,
    canSwitchBranch: Boolean,
    onSelect: (String) -> Unit,
    onNoteChange: (String) -> Unit,
    onSell: () -> Unit,
    onRetry: () -> Unit,
    onDismissError: () -> Unit,
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Box {
        KlinaraScreen(title = "Paket sat", modifier = modifier, onBack = onBack) {
            state.error?.let { ErrorBanner(message = it, retryLabel = "Kapat", onRetry = onDismissError) }

            when (val options = state.options) {
                Loadable.Loading ->
                    Text("Yükleniyor…", style = KlinaraType.bodyM, color = KlinaraTheme.colors.charcoalMuted)
                is Loadable.Failed ->
                    ErrorBanner(message = options.message, onRetry = if (options.isRetryable) onRetry else null)
                is Loadable.Loaded ->
                    if (options.value.isEmpty()) {
                        EmptyStateView(
                            title = "Satılabilir paket yok",
                            // Şubenin ADI yazılıyor: çok şubeli bir kiracıda "bu şubede"
                            // hangi şube olduğunu söylemiyordu.
                            message = emptyMessage(branchName, canSwitchBranch),
                            icon = Icons.Filled.ShoppingCart,
                        )
                    } else {
                        KlinaraCard(title = "Paket") {
                            KlinaraSearchablePicker(
                                options = options.value,
                                key = { it.id },
                                label = { it.name },
                                detail = { optionDetail(it) },
                                isSelected = { it.id == state.selectedId },
                                onSelect = { onSelect(it.id) },
                                searchLabel = "Paket ara",
                            )
                        }
                        state.selected?.let { SalePreview(it) }
                        KlinaraCard(title = "Not") {
                            KlinaraTextEditor(
                                label = "Satış notu",
                                value = state.note,
                                onValueChange = onNoteChange,
                                placeholder = "İsteğe bağlı",
                                error = state.fieldErrors["note"],
                            )
                        }
                        KlinaraButton(
                            title = "Paketi sat",
                            onClick = onSell,
                            icon = Icons.Filled.ShoppingCart,
                            isLoading = state.isSaving,
                            enabled = state.canSell,
                        )
                    }
            }
        }

        if (state.isSaving) AuthLoadingOverlay(message = "Satış kaydediliyor…")
    }
}

@Composable
private fun SalePreview(definition: PackageDefinition) {
    KlinaraCard(title = "Satış önizlemesi", footnote = previewFootnote(definition)) {
        definition.sortedItems.forEachIndexed { index, item ->
            if (index > 0) KlinaraDivider()
            KlinaraRow(
                label = item.serviceName,
                value = "${item.quantity} seans",
                detail = "${Money.format(item.unitListPriceMinor, definition.currency)} birim",
            )
        }
        KlinaraDivider()
        KlinaraRow(
            label = "Satış tutarı",
            value = Money.format(definition.totalPriceMinor, definition.currency),
            detail =
                definition.discountMinor?.let {
                    "Liste: ${Money.format(definition.listPriceMinor, definition.currency)} · " +
                        "İndirim: ${Money.format(it, definition.currency)}"
                },
        )
    }
}

private fun optionDetail(definition: PackageDefinition): String =
    "${definition.totalSessions} seans · ${Money.format(definition.totalPriceMinor, definition.currency)}"

private fun previewFootnote(definition: PackageDefinition): String =
    listOf(
        definition.validityDays?.let { "Satıştan itibaren $it gün geçerli" } ?: "Süresiz",
        if (definition.isTransferable) "Devredilebilir" else "Devredilemez",
        "Satış anındaki fiyat ve kalemler dondurulur; tanım sonradan değişse bile bu paket etkilenmez.",
    ).joinToString(" · ")

private fun emptyMessage(
    branchName: String?,
    canSwitchBranch: Boolean,
): String {
    if (branchName == null) return "Satışa açık paket tanımı bulunmuyor."
    val switchHint = if (canSwitchBranch) " Başka bir şubede satış için önce şubeyi değiştirin." else ""
    return "$branchName şubesinde satışa açık paket tanımı bulunmuyor.$switchHint"
}
